src/bigquery.ts

```diff
@@ -2125,8 +2126,42 @@ export class Geography {
  */
 export class BigQueryTimestamp {
   value: string;
-  constructor(value: Date | string) {
-    this.value = new Date(value).toJSON();
+  constructor(value: Date | PreciseDate | string | number) {
+    let pd: PreciseDate;
+    if (value instanceof PreciseDate) {
+      pd = value;
+    } else if (value instanceof Date) {
+      pd = new PreciseDate(value);
+    } else if (typeof value === 'string') {
+      if (/^\d{4}-\d{1,2}-\d{1,2}/.test(value)) {
+        pd = new PreciseDate(value);
+      } else {
+        const floatValue = Number.parseFloat(value);
```

Comment:
I'm vaguely concerned that the complexities of the type coercion here need a few more unit tests to catch all the corner cases (e.g. what happens if you pass NaN to timestamp()).

Reply:
Added an extra test for NaN and checking code coverage and all branches are covered now for that function.